Integrate coverage configuration and tests in CI pipeline

OlenaYefymenko opened the pull request:
This patch provides the coverage configuration, updates the project dependencies accordingly, and adds tests to the CI pipeline. This PR sets the stage for the next, where test functions will be integrated into the project.

Related to #18

OlenaYefymenko commented:
@webknjaz I've implemented the `coverage` configuration as we did in my previous project. However, you mentioned an option regarding this:

```
#!/usr/bin/env python
"""Django's command-line utility for administrative tasks."""
import os
import sys


def main():
    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "myproj.settings")

    # MyProject Customization: run coverage.py around tests automatically
    try:
        command = sys.argv[1]
    except IndexError:
        command = "help"

    running_tests = command == "test"
    if running_tests:
        from coverage import Coverage

        cov = Coverage()
        cov.erase()
        cov.start()

    try:
        from django.core.management import execute_from_command_line
    except ImportError as exc:
        raise ImportError(
            "Couldn't import Django. Are you sure it's installed and "
            "available on your PYTHONPATH environment variable? Did you "
            "forget to activate a virtual environment?"
        ) from exc
    execute_from_command_line(sys.argv)

    if running_tests:
        cov.stop()
        cov = Coverage()
        cov.save()
        covered = cov.report()
        if covered < 100:
            raise SystemExit(1)


if __name__ == "__main__":
    main()
```

I have doubts about whether I should keep the configuration as it is in the current PR or add it into `manage.py` using `contextlib.nullcontext()`. However, I've read that using a `coveragerc` file is preferable.

webknjaz commented:
Use the coverage file for the configuration but integrate nullcontext for enabling coverage collection conditionally. Also, use the contextmanager decorator to move that to a separate helper function.

OlenaYefymenko commented:
@webknjaz I've just added what we wrote in the live coding session.

webknjaz commented on .coveragerc:
```diff
@@ -0,0 +1,29 @@
+[report]
+skip_covered = true
+show_missing = true
+exclude_lines =
```
There's a new setting in coveragepy called `exclude_also`. There are default entries that coveragepy sets for `exclude_lines`. If you supply your own list, it replaces those entries, but with `exclude_also`, it extends the default entries with yours. It's considered better now. Also, in the future, it might be nice to integrate https://pypi.org/p/covdefaults which provides many popular settings automatically.
But for now, this small change should suffice:
```suggestion
exclude_also =
```

webknjaz commented on .coveragerc:
```diff
@@ -0,0 +1,29 @@
+dynamic_context = test_function
+parallel = true
+relative_files = true
+source =
+  tests
```
You don't have this top-level folder so it's probably unnecessary to use this setting. OTOH, it may be nice to move the tests to the repository root so perhaps keeping it as it is is warranted.

OlenaYefymenko replied:
Okay, I will create the directory in the next PR.

webknjaz replied:
Marking this as resolved.

webknjaz commented on .coveragerc:
```diff
@@ -0,0 +1,29 @@
+[html]
+show_contexts = true
+skip_covered = false
+
+[paths]
+source =
+  mysite
```
This section is not needed. It's used to map files in your repository with their site-packages copies. But since a Django project isn't `pip install`ed, there should be no need to configure this mapping.

webknjaz replied:
Marking this as resolved.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -6,6 +6,24 @@ on:  # yamllint disable-line rule:truthy
+        python-version: 3.11
+
+    - name: Install requirements
+      run: python -m pip install -r requirements/base.in
+           -c requirements/base.txt
```
Let's make the formatting more intuitive here:
```suggestion
      run: >-
        python -m pip install
        -r requirements/base.in
        -c requirements/base.txt
```

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -17,6 +35,10 @@ jobs:
         python-version: 3.11
 
+    - name: Install requirements
+      run: python -m pip install -r requirements/base.in
+           -c requirements/base.txt
```
```suggestion
      run: >-
        python -m pip install
        -r requirements/base.in
        -c requirements/base.txt
```

webknjaz commented on .pre-commit-config.yaml:
```diff
@@ -47,6 +47,7 @@ repos:
     - --django-settings-module=mysite.settings
     additional_dependencies:
     - pylint_django[with_django] >= 2.5.3
+    - coverage
```
It's usually nicer to have these entries sorted:
```suggestion
    - coverage
    - pylint_django[with_django] >= 2.5.3
```

webknjaz commented on manage.py:
```diff
@@ -1,9 +1,45 @@
-#!/usr/bin/env python
```
Don't remove this. It's unrelated to coverage and is necessary for running the script as `./manage.py`.

webknjaz replied:
Marking this as resolved.

webknjaz commented on manage.py:
```diff
@@ -1,9 +1,45 @@
-#!/usr/bin/env python
 """Django's command-line utility for administrative tasks."""
-import os
 import sys
+import os
```
Keep the entries sorted. Also, making formatting changes in the same PR as functional changes is considered a very bad tone because this makes it _much harder_ to compare the versions during review.
```suggestion
import os
import sys
```

webknjaz replied:
Marking this as resolved.

webknjaz commented on requirements/base.in:
```diff
@@ -1 +1,2 @@
 Django ~= 3.0
+coverage==7.3.2
```
Let's not add test-only dependencies to the base requirements. Use a separate requirements file for non-production envs. Also, no need to pin the version in this file because it's an input file and the pins need to be updateable by pip-tools in the future, don't prevent this.

OlenaYefymenko replied:
ок

webknjaz replied:
Marking this as resolved.

webknjaz commented on requirements/base.txt:
Per above, this lockfile shouldn't be updated.

webknjaz replied:
Here.

OlenaYefymenko replied:
> Per above, this lockfile shouldn't be updated.

Ok, but in this PR, I should separate the `coverage` into another section of the dep. That's why the `base.txt` file also needs updating after this. Maybe we can keep it this way?

webknjaz replied:
No, it shouldn't be updated. You already reverted most of the changes except for comment lines with file paths.

webknjaz replied:
Marking this as resolved.

webknjaz commented on manage.py:
```diff
@@ -1,9 +1,45 @@
+    with cov.collect():
+        yield
+    cov.save()
+    print(f'Coverage is {cov.report()}%')  # xml, html
```
The comment is pointless here, because it doesn't mean anything, nor is related to printing out the text on this line.
```suggestion
    print(f'Coverage is {cov.report()}%')
```

OlenaYefymenko replied:
Agree

webknjaz commented on manage.py:
```diff
@@ -1,9 +1,45 @@
+    Uses coverage package to collect coverage, generate reports
+    and print coverage percentage.
+    """
+    # pylint: disable-next=import-outside-toplevel
+    from coverage import Coverage
```
It's possible to update `.pylintrc` instead of making the inline pylint-disable comment:
```ini
allow-any-import-level =
  coverage,
```
But it's up to you, keeping it here is also fine.

OlenaYefymenko replied:
Ok, I will change it. Thanks.

webknjaz replied:
Marking this as resolved.

webknjaz commented on manage.py:
```diff
@@ -1,9 +1,45 @@
-import os
 import sys
+import os
+from contextlib import nullcontext, contextmanager
```
An alternative solution to importing coverage is
```suggestion
from contextlib import contextmanager, nullcontext, suppress

with suppress(ImportError):
    from coverage import Coverage
```
This way, it wouldn't be imported in the function context, which may be perceived as a bit more elegant.
Though, the downside is that the import would be attempted even when unneeded (for non-test subcommands).

OlenaYefymenko commented:
Sorry, in next PR I will add  `# yamllint disable-line rule:truthy`.

webknjaz commented:
I got some notification regarding a "truthy" yamllint rule and can't find that comment to see the context here. What was that about? Can you link it?

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -6,6 +6,32 @@ on:  # yamllint disable-line rule:truthy
+      with:
+        python-version: 3.11
+
+    - name: Install requirements
+      run: >-
+        python -m pip install
+        -r requirements/base.in
+        -c requirements/base.txt
```
These shouldn't be separate, otherwise, they'll be overwritten by the following `pip install` with possibly different versions of things.

OlenaYefymenko replied:
> These shouldn't be separate, otherwise, they'll be overwritten by the following `pip install` with possibly different versions of things.

I'm sorry, but I don't clearly understand how the installation of requirements should look in the configuration file. Here is your previous fix https://github.com/kpi-web-guild/django-girls-blog-OlenaEfymenko/pull/31#discussion_r1425681784. 
I'll try to find some examples and docs, maybe.

OlenaYefymenko replied:
`run: python -m pip install -r requirements/base.in -c requirements/base.txt`
like this variant

webknjaz replied:
You mustn't install base explicitly/directly.

OlenaYefymenko replied:
Yes, I understood. Thank you)

webknjaz replied:
> Here is your previous fix [#31 (comment)](https://github.com/kpi-web-guild/django-girls-blog-OlenaEfymenko/pull/31#discussion_r1425681784).

That was just an example of a formatting change, orthogonal to the fact that the step shouldn't be included, actually.

webknjaz replied:
Marking this as resolved.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -17,6 +43,18 @@ jobs:
       with:
         python-version: 3.11
 
+    - name: Install requirements
+      run: >-
+        python -m pip install
+        -r requirements/base.in
+        -c requirements/base.txt
```
Same here: these versions must be consistent with the other constraints file.

webknjaz replied:
Marking this as resolved.

webknjaz commented on requirements/base.in:
```diff
@@ -1 +1 @@
-Django ~= 3.0
```
How is this change related to the PR?

OlenaYefymenko replied:
I remember about atomicity, but in this PR I added `coverage` and restructured requirements. It may be complicated to separate them.

webknjaz replied:
This file just doesn't need any changes at all. No need to separate anything from anywhere. You add other deps into a different lockfile.

webknjaz replied:
Marking this as resolved.

webknjaz commented on requirements/base.txt:
```diff
@@ -2,12 +2,12 @@
 # by the following command:
 #
-#    pip-compile base.in
+#    pip-compile requirements/base.in
```
Looks like you're running `pip-tools` from a different directory, then before. Be consistent so that it doesn't generate pointless changes.

webknjaz replied:
Adding a 👍 doesn't count as reverting an unrelated change.

OlenaYefymenko replied:
> Adding a 👍 doesn't count as reverting an unrelated change.

Yes, it looks like this pull request needs a lot of action with fixing commits.

webknjaz replied:
Marking this as resolved.

webknjaz commented on requirements/dev.in:
Don't forget to include the base requirements so that the pins are consistent and don't conflict with each other.

OlenaYefymenko replied:
Ok

webknjaz replied:
Add the input file, not the constraints one.

webknjaz replied:
Marking this as resolved.

OlenaYefymenko commented:
> I got some notification regarding a "truthy" yamllint rule and can't find that comment to see the context here. What was that about? Can you link it?

Yes, first I pushed without this comment and CI linters crashed, but then I fixed this and deleted the comment

webknjaz commented:
It's best to communicate what's happening. Deleting things only complicates everything.

webknjaz reviewed (changes requested):
Reminding to undo reshuffling the imports and address other comments.

OlenaYefymenko commented:
> Reminding to undo reshuffling the imports and address other comments.

I've merged the commits with the import into the previous, resolved conflicts. I hope this is appropriate for this issue. Also, I assume I've checked all comments, except for this one: https://github.com/kpi-web-guild/django-girls-blog-OlenaEfymenko/pull/31#discussion_r1425699563
Can we postpone addressing that one?

webknjaz commented on .pre-commit-config.yaml:
```diff
@@ -46,8 +46,10 @@ repos:
+    - coverage
     - pylint_django[with_django] >= 2.5.3
 
+
```
Here.

OlenaYefymenko replied:
I assume that we are integrating `coverage` into the project within the current PR. Additionally, in this commit, we should modify the `pre-commit tool` by adding `coverage`. Sorry, I'm a bit confused.

webknjaz replied:
Exactly. Not formatting changes in disconnected places.

webknjaz replied:
Marking this as resolved.

webknjaz commented on manage.py:
```diff
@@ -1,9 +1,47 @@
 #!/usr/bin/env python
+
```
Here.

OlenaYefymenko replied:
Thank you. I've squashed commits.

webknjaz replied:
Okay, now remove the formatting changes.

OlenaYefymenko replied:
I'm sorry, yes, I assume when I resolved conflicts I didn't finish this issue with `#!/usr/bin/env python`.
Also, I didn't notice just the empty line in the config. At first, I guess you meant about `coverage`. Now I hope the PR is atomic. I understand that `pylint` was added in the dependencies, but it's just one minor change not closely related to this PR.

webknjaz replied:
Marking this as resolved.

webknjaz reviewed (changes requested):
These are some unrelated changes I spotted.
